homeassistant/components/ihc/__init__.py

```diff
@@ -98,35 +168,76 @@
 
 
 def setup(hass, config):
+    """Set up the IHC platform."""
+    conf = config.get(DOMAIN)
+    for index, controller_conf in enumerate(conf):
+        if not ihc_setup(hass, config, controller_conf, index):
+            return False
+
+    return True
+
+
+def ihc_setup(hass, config, conf, controller_id):
     """Set up the IHC component."""
     from ihcsdk.ihccontroller import IHCController
-    conf = config[DOMAIN]
+
     url = conf[CONF_URL]
     username = conf[CONF_USERNAME]
     password = conf[CONF_PASSWORD]
-    ihc_controller = IHCController(url, username, password)
 
+    ihc_controller = IHCController(url, username, password)
     if not ihc_controller.authenticate():
         _LOGGER.error("Unable to authenticate on IHC controller")
         return False
 
     if (conf[CONF_AUTOSETUP] and
-            not autosetup_ihc_products(hass, config, ihc_controller)):
+            not autosetup_ihc_products(hass, config, ihc_controller,
+                                       controller_id)):
         return False
-
-    hass.data[IHC_DATA] = {
+    # Manual configuration
+    get_manual_configuration(hass, config, conf, ihc_controller,
+                             controller_id)
+    # Store controler configuration
+    ihc_key = IHC_DATA.format(controller_id)
+    hass.data[ihc_key] = {
         IHC_CONTROLLER: ihc_controller,
         IHC_INFO: conf[CONF_INFO]}
-
     setup_service_functions(hass, ihc_controller)
     return True
 
 
-def autosetup_ihc_products(hass: HomeAssistantType, config, ihc_controller):
+def get_manual_configuration(hass, config, conf, ihc_controller,
+                             controller_id):
+    """Get manual configuration IHC devices."""
+    for component in IHC_PLATFORMS:
+        if component in conf:
+            component_setup = conf.get(component)
+            for sensor_id, sensor_cfg in component_setup.items():
+                device = {
+                    'ihc_id': sensor_id,
+                    'ctrl_id': controller_id,
+                    'product': {
+                        'name': sensor_cfg[CONF_NAME],
+                        'note': sensor_cfg.get(CONF_NOTE) or '',
+                        'position': sensor_cfg.get(CONF_POSITION) or ''},
+                    'product_cfg': {
+                        'type': sensor_cfg.get(CONF_TYPE),
+                        'inverting': sensor_cfg.get(CONF_INVERTING),
+                        'dimmable': sensor_cfg.get(CONF_DIMMABLE),
+                        'unit': sensor_cfg.get(CONF_UNIT_OF_MEASUREMENT)
+                    }
+                }
+                discovery_info = {sensor_cfg[CONF_NAME]: device}
+                discovery.load_platform(hass, component, DOMAIN,
```

Comment:
Only call `load_platform` once per component. Collect all the discovery_info in a container and pass that as discovery_info. Iterate the container in the platform to create all entities. Then add all entities in one call to `add_entities`.